### PR TITLE
Upgrade flysystem to version ^2.0

### PR DESCRIPTION
**Pull request by engcom-Kilo**
### Related Pull Requests
https://github.com/magento/partners-magento2ee/pull/530

### Fixed Issues (if relevant)

Resolves https://github.com/magento/magento2/issues/31177

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by engcom-Kilo** [comment]
@magento run all tests

**Comment by engcom-Kilo** [comment]
@magento run all tests

**Comment by engcom-Kilo** [comment]
@magento run all tests

**Comment by engcom-Kilo** [comment]
@magento run all tests

**Comment by engcom-Kilo** [comment]
@magento run all tests

**Comment by engcom-Kilo** [comment]
@magento run all tests

**Comment by engcom-Kilo** [comment]
@magento run all tests

File: app/code/Magento/AwsS3/Driver/AwsS3.php [hunk 1]
```diff
@@ -89,8 +88,8 @@ public function test(): void
     {
         try {
             $this->adapter->write(self::TEST_FLAG, '', new Config(self::CONFIG));
-        } catch (Exception $exception) {
-            throw new DriverException(__($exception->getMessage()), $exception);
+        } catch (\League\Flysystem\FilesystemException $e) {
```
Comment (sivaschenko) [inline app/code/Magento/AwsS3/Driver/AwsS3.php]: I believe it's better to catch `Exception` as well, to be sure (applicable for all other catch blocks in this class)

Please move `League\Flysystem\FilesystemException` to the `use` block

File: app/code/Magento/AwsS3/Model/Cached/CachedAdapter.php [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+
+/**
+ * Aws cached adapter model.
+ */
+class CachedAdapter implements FilesystemAdapter
+{
+    /**
+     * @var FilesystemAdapter
+     */
+    private FilesystemAdapter $adapter;
```
Comment (sivaschenko) [inline app/code/Magento/AwsS3/Model/Cached/CachedAdapter.php]: We should still support PHP 7.3. I don't think it supports parameter types

File: app/code/Magento/AwsS3/Model/Cached/CachedAdapter.php [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+        $object = $result->jsonSerialize() + compact('path');
+        $this->cache->updateObject($path, $object, true);
+
+        return $result;
+    }
+
+    /**
+     * Get file metadata.
+     *
+     * @deplacated There is no getMetadata() method in FilesystemAdapter anymore.
```
Comment (sivaschenko) [inline app/code/Magento/AwsS3/Model/Cached/CachedAdapter.php:280]: Do we need to introduce this deplacated method?

File: app/code/Magento/RemoteStorage/Driver/Cache/CacheInterface.php [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+declare(strict_types=1);
+
+namespace Magento\RemoteStorage\Driver\Cache;
+
+use League\Flysystem\FilesystemReader;
+
+/**
+ * Remote storage cache interface.
+ */
+interface CacheInterface extends FilesystemReader
```
Comment (sivaschenko) [inline app/code/Magento/RemoteStorage/Driver/Cache/CacheInterface.php]: I wouldn't introduce extension points (interfaces) if that is not necessary

File: app/code/Magento/RemoteStorage/Model/GetPathInfo.php [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+        $path = rtrim($path, $separators);
+
+        $basename = preg_replace('#.*?([^' . preg_quote($separators, '#') . ']+$)#', '$1', $path);
+
+        if (DIRECTORY_SEPARATOR === '/') {
+            return $basename;
+        }
+        // @codeCoverageIgnoreStart
+        // Extra Windows path munging. This is tested via AppVeyor, but code
```
Comment (sivaschenko) [inline app/code/Magento/RemoteStorage/Model/GetPathInfo.php]: We do not need to support Windows OS.

**Review by sivaschenko — COMMENTED** [review]
Thanks for the pull request! Please see my comments

**Comment by engcom-Kilo** [comment]
@magento run all tests

**Comment by engcom-Kilo** [comment]
@magento run all tests

File: app/code/Magento/AwsS3/Model/Cached/CachedAdapter.php (hunk 2 above)
Comment (engcom-Kilo) [reply app/code/Magento/AwsS3/Model/Cached/CachedAdapter.php]: Removed.

File: app/code/Magento/RemoteStorage/Driver/Cache/CacheInterface.php (hunk 4 above)
Comment (engcom-Kilo) [reply app/code/Magento/RemoteStorage/Driver/Cache/CacheInterface.php]: Moved Reader interface methods to Cache interface. Removed inheritance

**Comment by engcom-Kilo** [comment]
@magento run all tests

**Comment by engcom-Kilo** [comment]
@magento run all tests

File: app/code/Magento/AwsS3/Model/Cached/CachedAdapter.php (hunk 3 above)
Comment (engcom-Kilo) [reply app/code/Magento/AwsS3/Model/Cached/CachedAdapter.php:280]: There are several clients for this method:
Magento\Eav\Model\Attribute\Data\Image, Magento\MediaGallerySynchronization\Model\CreateAssetFromFile, Magento\Framework\File\Mime.

File: app/code/Magento/AwsS3/Driver/AwsS3.php (hunk 1 above)
Comment (engcom-Kilo) [reply app/code/Magento/AwsS3/Driver/AwsS3.php]: Done.

**Comment by engcom-Kilo** [comment]
@magento run all tests

File: app/code/Magento/RemoteStorage/Model/GetPathInfo.php (hunk 5 above)
Comment (engcom-Kilo) [reply app/code/Magento/RemoteStorage/Model/GetPathInfo.php]: Removed

File: app/code/Magento/RemoteStorage/composer.json [hunk 6]
```diff
@@ -3,7 +3,8 @@
     "description": "N/A",
     "require": {
         "php": "~7.3.0||~7.4.0",
-        "magento/framework": "*"
+        "magento/framework": "*",
+        "predis/predis": "*"
```
Comment (sivaschenko) [inline app/code/Magento/RemoteStorage/composer.json:7]:
```suggestion
        "predis/predis": "^1.1"
```

File: app/code/Magento/RemoteStorage/composer.json [hunk 7]
```diff
@@ -14,8 +15,7 @@
         "magento/module-media-storage": "*",
         "magento/module-import-export": "*",
         "magento/module-catalog-import-export": "*",
-        "magento/module-downloadable-import-export": "*",
-        "predis/predis": "*"
+        "magento/module-downloadable-import-export": "*"
```
Comment (sivaschenko) [inline app/code/Magento/RemoteStorage/composer.json:18]: Documentation should be updated as there is no need to require `predis/predis` package https://devdocs.magento.com/guides/v2.4/config-guide/remote-storage/config-remote-storage-caching.html

File: app/code/Magento/RemoteStorage/Driver/Cache/CacheInterface.php [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+use League\Flysystem\FilesystemException;
+use League\Flysystem\StorageAttributes;
+use League\Flysystem\UnableToCheckFileExistence;
+use League\Flysystem\UnableToReadFile;
+use League\Flysystem\UnableToRetrieveMetadata;
+
+/**
+ * Remote storage cache interface.
+ */
+interface CacheInterface
```
Comment (sivaschenko) [inline app/code/Magento/RemoteStorage/Driver/Cache/CacheInterface.php:19]: The interface should be removed. There is only one implementation and I wouldn't introduce a new extension point in the patch release

File: app/code/Magento/AwsS3/Model/Cached/CachedAdapter.php (hunk 3 above)
Comment (sivaschenko) [reply app/code/Magento/AwsS3/Model/Cached/CachedAdapter.php:280]: The comment and `@deplacated` annotation should be removed.

Is implementing `League\Flysystem\FilesystemAdapter` necessary?

File: app/code/Magento/RemoteStorage/Model/Storage/Handler/Predis.php [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+
+use Magento\Framework\Serialize\Serializer\Json;
+use Magento\RemoteStorage\Model\Storage\CacheStorage;
+use Magento\RemoteStorage\Model\Storage\GetCleanedContents;
+use Predis\Client;
+
+/**
+ * Redis cache model.
+ */
+class Predis implements CacheStorageHandlerInterface
```
Comment (sivaschenko) [inline app/code/Magento/RemoteStorage/Model/Storage/Handler/Predis.php:18]: Introduced classes should be covered by integration tests

**Review by sivaschenko — CHANGES_REQUESTED** [review]

**Comment by sivaschenko** [comment]
Flysystem is now updated in 2.4-develop

**Pull request closed without merging**